File: go/bob/bob.go
package bob

import (
	"fmt"
	"strings"
)

func Hey(remark string) string {
	trimmed := strings.TrimSpace(remark)

	if trimmed == "" {
		return "Fine. Be that way!"
	} else if isQuestion(trimmed) && isAllCaps(trimmed) && hasLetters(trimmed) {
		return "Calm down, I know what I'm doing!"
	} else if isQuestion(trimmed) {
		return "Sure."
	} else if isAllCaps(trimmed) && hasLetters(trimmed) {
		return "Whoa, chill out!"
	} else {
		return "Whatever."
	}
}

func isQuestion(value string) bool {
	return strings.HasSuffix(value, "?")
}

func isAllCaps(value string) bool {
	return value == strings.ToUpper(value)
}

func hasLetters(value string) bool {
	allLetters := "abcdefghijklmnopqrstuvwxyz"
	return strings.ContainsAny(value, fmt.Sprintf("%s%s", allLetters, strings.ToUpper(allLetters)))
}

File: swift/bob/Sources/Bob/Bob.swift
struct Bob {
    
    static func hey(_ input: String) -> String {
        
        if input.isOnlyWhitespace {
            return "Fine. Be that way!"
        }
        
        if input.isAllCaps {
            return "Whoa, chill out!"
        }
        
        if input.isQuestion {
            return "Sure."
        }
        
        return "Whatever."
    }
}

private extension String {
    
    var isOnlyWhitespace: Bool {
        return self.filter { !$0.isWhitespace }.isEmpty
    }
    
    var isQuestion: Bool {
        return self.endsWith("?")
    }
    
    var isAllCaps: Bool {
        let letters = self.filter { $0.isLetter }
        
        if letters.isEmpty {
            return false
        }
        
        return letters.filter { $0.isLowercase }.isEmpty
    }
    
    func endsWith(_ char: Character) -> Bool {
        guard let lastChar = self.last else {
            return false
        }
        
        return char == lastChar
    }
}
